src/rpc/client.cpp

```diff
@@ -15,118 +16,245 @@
+                return std::string("Error parsing JSON:") + err.unparsedArg;
+            }
+        }, failure);
+}
+
+optional<pair<vector<bool>, vector<bool>>> paramsToConvertFor(const string& method) {
     auto search = rpcCvtTable.find(method);
     if (search != rpcCvtTable.end()) {
         return search->second;
     } else {
-        return set<int>();
+        return nullopt;
     }
 }
 
-/** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null)
- * as well as objects and arrays.
- */
-UniValue ParseNonRFCJSONValue(const string& strVal)
+optional<UniValue> ParseNonRFCJSONValue(const string& strVal)
 {
     UniValue jVal;
-    if (!jVal.read(string("[")+strVal+string("]")) ||
-        !jVal.isArray() || jVal.size()!=1)
-        throw runtime_error(string("Error parsing JSON:")+strVal);
-    return jVal[0];
+    if (jVal.read(string("[")+strVal+string("]")) && jVal.isArray() && jVal.size() == 1) {
+        return jVal[0];
+    } else {
+        return nullopt;
+    }
 }
 
-/** Convert strings to command-specific RPC representation */
-UniValue RPCConvertValues(const string &strMethod, const vector<string> &strParams)
+tl::expected<UniValue, ConversionFailure>
+RPCConvertValues(const string &method, const vector<string> &strArgs)
 {
     UniValue params(UniValue::VARR);
+    auto paramsToConvert = paramsToConvertFor(method);
+    vector<bool> requiredParams{};
+    vector<bool> optionalParams{};
+    if (paramsToConvert.has_value()) {
+        requiredParams = paramsToConvert.value().first;
+        optionalParams = paramsToConvert.value().second;
+    } else {
+        return tl::expected<UniValue, ConversionFailure>(tl::unexpect, UnknownRPCMethod());
+    }
+
+    if (strArgs.size() < requiredParams.size()
+        || requiredParams.size() + optionalParams.size() < strArgs.size()) {
+        return tl::expected<UniValue, ConversionFailure>(
+                tl::unexpect,
+                WrongNumberOfArguments(requiredParams.size(), optionalParams.size(), strArgs.size()));
+    }
+
+    vector<bool> allParams(requiredParams.begin(), requiredParams.end());
+    allParams.reserve(requiredParams.size() + optionalParams.size());
+    allParams.insert(allParams.end(), optionalParams.begin(), optionalParams.end());
 
-    auto paramsToConvert = paramsToConvertFor(strMethod);
-    for (vector<int>::size_type idx = 0; idx < strParams.size(); idx++) {
-        const string& strVal = strParams[idx];
+    for (vector<int>::size_type idx = 0; idx < strArgs.size(); idx++) {
+        const bool shouldConvert = allParams[idx];
+        const string& strVal = strArgs[idx];
 
-        params.push_back(
-                paramsToConvert.count(idx) != 0
-                // parse string as JSON, insert bool/number/object/etc. value
-                ? ParseNonRFCJSONValue(strVal)
-                // insert string value directly
-                : strVal);
+        if (shouldConvert) {
```

Comment:
I see that this ternary is removed in this later commit. If the original conditional were preserved, this change would also have been smaller (instead leading with `if (!shouldConvert) {`). I'd personally prefer this refactor for commit diff clarity.